pkg/util/globalerror/grpc.go

```diff
@@ -82,47 +81,31 @@ func (e *grpcContextError) GRPCStatus() *grpcstatus.Status {
 // and grpc/status packages, but which preserve the original error
 // semantics.
 type ErrorWithStatus struct {
-	UnderlyingErr error
-	Status        *status.Status
+	Causes []error
+	Status *status.Status
 }
 
-func NewErrorWithGRPCStatus(originalErr error, code codes.Code, details *mimirpb.ErrorDetails) ErrorWithStatus {
-	stat := status.New(code, originalErr.Error())
-	if details != nil {
-		if statWithDetails, err := stat.WithDetails(details); err == nil {
-			return ErrorWithStatus{
-				UnderlyingErr: originalErr,
-				Status:        statWithDetails,
-			}
-		}
-	}
-	return ErrorWithStatus{
-		UnderlyingErr: originalErr,
-		Status:        stat,
-	}
-}
-
-func (e ErrorWithStatus) Error() string {
+func (e *ErrorWithStatus) Error() string {
 	return e.Status.Message()
 }
 
-func (e ErrorWithStatus) Unwrap() error {
-	return e.UnderlyingErr
+func (e *ErrorWithStatus) Unwrap() []error {
```

Comment:
_not very actionable, but decided to bring this up anyways_

this changes the contract of this type. From what I could find implementing `Unwrap() []error` will still work with `errors.Is()` ([code](https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/errors/wrap.go;l=44-79)), but not with `errors.Unwrap()` ([code](https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/errors/wrap.go;l=17-25)). I think that should be fine because we don't use `errors.Unwrap()` in the Mimir repo.

Reply:
Hi @dimitarvdimitrov, 
This is what the `errors` package [documentation](https://pkg.go.dev/errors) states:

An error e wraps another error if e's type has one of the methods
```
Unwrap() error
Unwrap() []error
```

So it is enough to implement one of these two in order to have correct error wrapping. Moreover, by doing it this way we can remove the `Is()` and `As()` functions that the `grpcContextError` type implemented.